### PR TITLE
[flutter] unify reassemble and fast reassemble

### LINKED ISSUE #84347: Flutter Hot reload/Hot restart gets stuck on reassemble in callbackDispatcher isolate
Edit:

If you're reading this because the issue was referenced and you can't get hot reload to work, here's a TL;DR:

This happens when using fvm < 2.2.0. To resolve the issues with hot reload, upgrade your fvm installation.

Original Post:

I have a problem with my Flutter project, which causes Hot Reload / Hot Restart to hang indefinitely sometimes.
It happens on Android and iOS devices, real and simulator/emulator.
It does not happen on every time I use hot reload/hot restart, but after a few times it usually gets triggered fairly reliably. 

Here a gist of the debug output of flutter attach --verbose:

https://gist.github.com/knaeckeKami/46970ac3073ff38d367f9ff49cc98322

Relevant lines:

```
[        ] Performing hot reload...
[  +38 ms] Scanned through 1349 files in 29ms
[        ] Syncing files to device iPhone 12 Pro Max...
[        ] Compiling dart to kernel with 1 updated files
[   +5 ms] <- recompile package:neocortex_mobile/main.staging.dart be2a7fa8-087d-4693-a996-2cd6d1b763f1
[        ] package:neocortex_mobile/matrix/chat.dart
[        ] <- be2a7fa8-087d-4693-a996-2cd6d1b763f1
[ +233 ms] Updating files.
[  +14 ms] DevFS: Sync finished
[   +1 ms] Syncing files to device iPhone 12 Pro Max... (completed in 254ms)
[        ] Synced 0.1MB.
[ +275 ms] <- accept
[        ] reloaded 1 of 2938 libraries
[  +70 ms] Reassembling application
[+2006 ms] This is taking a long time; will now check for paused isolates.
[  +90 ms] Found 0 newly paused isolate(s).
```

After the line 

` Found 0 loadPausedIsolatesFound newly paused isolate(s).` 

the tools just hangs and does not react to any commands. 
After quitting the tool with STRG-C and re-attaching to the running flutter app, I can hot reload/ hot restart again.

After digging a bit, I found that the tool gets stuck on this line:

https://github.com/flutter/flutter/blob/07921b561d6f79a5cfe0401b1b081a5d581bd771/packages/flutter_tools/lib/src/run_hot.dart#L1265

The hot reload actually works i.e. I see the changes that I have made deployed on the device, it's just that the tool keeps hanging here.
I tried commenting out this line, which made hot reload and hot restart work again reliably without hanging (but I don't know about possible side effects that this causes).

I added more debug logging to this file to get an idea of what's going on. Here's a gist of the output of run_hot.dart with more logging added:

https://gist.github.com/knaeckeKami/566f920560a351197ff1d56b72add98b

The modified run_hot.dart with additional logging is here:
https://gist.github.com/knaeckeKami/af9cbd138ef55c1f2e8a3dd45ec821a0 

My interpretation of this (I don't really know what's going on here): 
- There is an isolate called main, for which the reassembleFuture completes.
See the lines
```
[  +49 ms] found isolate main isolates/1112832209456463 [Event kind: Resume, timestamp: 1623258672622]
```
and
```
[   +2 ms] finished reassemble for isolates/1112832209456463
```
- There is another isolate called callbackDispatcher, where the reassambleFuture never completes. (no lines "finished reassemble for... printed for this isolate)

## Steps to Reproduce

Unfortunately, I am not able to produce a minimal example of this. It only happens in a big-ish production app.

I don't know much about the internals of the hot reload/ hot restart system, so I would not know how to start trying to reproduce this. Obviously, this works with a newly created flutter app. But I'm happy to dig deeper when someone with more insight gives me a few pointers.

```
 fvm flutter doctor --verbose
Building flutter tool...
[✓] Flutter (Channel unknown, 2.2.1, on macOS 11.4 20F71 darwin-x64, locale en-GB)
    • Flutter version 2.2.1 at /Users/kami/fvm/versions/2.2.1
    • Framework revision 02c026b03c (2 weeks ago), 2021-05-27 12:24:44 -0700
    • Engine revision 0fdb562ac8
    • Dart version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/kami/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /Users/kami/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.1)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • 1C848851-B380-4185-B9D5-54C06C495E9D • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • macOS 11.4 20F71 darwin-x64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 91.0.4472.101

• No issues found!
```

</details>

Same results happen also on the current master channel.

### LINKED ISSUE #84357: Changes to widget hierarchy doesn't require reassemble of RenderObject tree
When performing a hot reload, the framework will reassemble both the element tree (and their widgets/state objects), as well as the RenderObject tree, triggering repaint/semantics/layout. The latter isn't necessary unless there was a code change in a render object. Skipping the reassemble of render objects for a small change on a Pixel 4 decreased hot reload time from 650 ms to 575 ms.

Similarly, if there was only a code change in a RenderObject, then the widget hierarchy doesn't need to be reassembled at all.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Modifies `fastReassemble` to a be a special case of regular `reassemble`, which should remove the potential for large performance deltas.

Now that the new invalidations strategy is the default, fast reassemble should be triggered more often since it requires a single library delta. This adds special cases to regular reassemble to:

1. Skip reassembling Elements until it reaches the first element that matches.
2. Skip RenderObject reassemble if we know that change was from a class member (fixes https://github.com/flutter/flutter/issues/84357 )

After this change, we should enable by default on master/dev/beta, along with some communication on changing behavior. That should give us enough information to make a call on whether this can be the default, or whether we should pull this out entirely.

Also fixes https://github.com/flutter/flutter/issues/84347 , which is caused by fast reassemble not working well on background isolates with a flutter view (but no UI)

### CODE REVIEW

**Comment by jonahwilliams** [comment]
This changes the API of reassemble in order to pass the config through, but leaves it the same for Elements/State objects since users / widget library authors may have made their own implementations here

**Comment by jonahwilliams** [comment]
Still adding more tests

**Comment by jonahwilliams** [comment]
tests added

File: packages/flutter/lib/src/widgets/binding.dart [hunk 1]
```diff
@@ -444,16 +444,12 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
...
+          BindingBase.debugReassembleConfig = DebugReassembleConfig(widgetName: className);
+          try {
+            await reassembleApplication();
+          } finally {
+            BindingBase.debugReassembleConfig = null;
           }
-          if (renderViewElement != null) {
-            markElementsDirty(renderViewElement!);
-          }
-          await endOfFrame;
```
Comment (jmagman) [inline packages/flutter/lib/src/widgets/binding.dart:456]: Did you mean to remove this `await`?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/binding.dart:456]: Yes, this method is intended to mirror https://github.com/flutter/flutter/blob/85df73d0e34e1a80d3e468303926434deb790ef9/packages/flutter/lib/src/foundation/binding.dart#L165-L169 but with an extra config

Will add a comment

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/binding.dart:456]: comment added

**Review by jmagman — COMMENTED** [review]
The code itself LGTM but I'd like for @zanderso or someone more familiar with this to take a look.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -3049,6 +3050,7 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
       _widget = widget;
 
   Element? _parent;
+  DebugReassembleConfig? _debugReassembleConfig;
```
Comment (zanderso) [inline packages/flutter/lib/src/widgets/framework.dart:3053]: Does adding this additional field to `Element` have any effect on release-mode memory footprint? Asked a different way, will this field be tree-shaken in a release build?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:3053]: This field is tree shaken, I inspected the size analysis for hello_world and it was not retained

**Review by zanderso — APPROVED** [review]

File: packages/flutter/lib/src/foundation/binding.dart [hunk 3]
```diff
@@ -629,3 +636,23 @@ abstract class BindingBase {
...
+  DebugReassembleConfig({
+    this.widgetName,
+  }) {
+    if (!kDebugMode) {
+      throw FlutterError('Cannot instaniate DebugReassembleConfig in profile or release mode.');
+    }
+  }
+
+  /// The name of the widget that was modified, or `null` if the change was elsewhere.
+  final String? widgetName;
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/binding.dart:657]: This layer isn't supposed to know about widgets. Can we move this logic to the widgets layer?

Comment (Hixie) [reply packages/flutter/lib/src/foundation/binding.dart:657]: Looks like the only use at lower layers is to disable the rendering layer's reassemble logic? Maybe we can just skip calling the reassemble logic in the rendering layer and just make fastReassemble invoke the relevant bits (like resetting the error count) and directly reassemble the widget tree with the class name? I'll file a bug to track this.

**Pull request merged**